Passport

JadziaHax opened the pull request.

Commit:
Add modules for passport.js and JSON Web Tokens

Installs 'passport', 'passport-jwt', and 'jsonwebtoken' npm modules as
dependencies in package.json

Commit:
Add passport authorization middleware, login, rename sessions route file

sessions.js in routes is now called authenticate.js. Auth system files
now live in src/config/auth. POSTing to authenticate will now check a
submitted username and password against the database and determine if a
match for both is found, responding with a token if so, otherwise
telling the user that they have submitted an invalid combination.

Commit:
Correct module paths in app.js.

Commit:
Add missing equal sign on line 17 of authenticate.js in routes.

Commit:
Correct passport module import syntax on line 12 of app.js.

Commit:
Make server respond to authentication requests with token on valid login

POST requests to /authenticate will now be checked for an email and
password that are compared against the database. If a user record with
the given email is found, the hashed password from that record is
compared against a hashed version of the POSTed password and if it
matches a token is returned in the response. If the email is not found
in the user records or the hash password submission does not match the
hashed password from the database, no token is generated or returned.

andioneto commented on server/app.js:
```diff
@@ -22,9 +25,9 @@ app.use(cookieParser());
+app.use('/authenticate', authenticate);
 
-app.use('*', (req, res, next) => {
+app.use('/', (req, res, next) => {
```
We might have to change this back to a wildcard (*), I'll try it after this merges and see what happens.

Commit:
Authorization for actions to /users; Move registration to /register

Registration now happens by POSTing user registration form data to
/resgistration. Passport.js authorization now happens before every
action at /users.
